pkg/watcher/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -63,26 +65,40 @@ func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
 		return err
 	}
 
-	if r.cfg.GetDisableAnnotationUpdate() {
-		// Don't update any annotations - nothing else to do.
-		return nil
+	if a := pr.GetAnnotations(); !r.cfg.GetDisableAnnotationUpdate() && (result.GetName() != a[annotation.Result] || record.GetName() != a[annotation.Record]) {
+		// Update PipelineRun with Result Annotations.
+		patch, err := annotation.Add(result.GetName(), record.GetName())
+		if err != nil {
+			log.Errorf("error adding Result annotations: %v", err)
+			return err
+		}
+		pr, err = r.pipelineclientset.TektonV1beta1().PipelineRuns(pr.GetNamespace()).Patch(pr.Name, types.MergePatchType, patch)
+		if err != nil {
+			log.Errorf("PipelineRun.Patch: %v", err)
+			return err
+		}
+	} else {
+		log.Infof("skipping CRD patch - annotation patching disabled [%t] or annotations already match", r.cfg.GetDisableAnnotationUpdate())
 	}
 
-	if a := pr.GetAnnotations(); result.GetName() == a[annotation.Result] && record.GetName() == a[annotation.Record] {
-		// Result annotations are already present in the PipelineRun.
-		// Nothing else to do.
-		return nil
-	}
+	// If the PipelineRun is complete and not yet marked for deletion, cleanup the run resource from the cluster.
+	if pr.IsDone() && r.cfg.GetCompletedResourceGracePeriod() != 0 {
+		if o := pr.GetOwnerReferences(); len(o) > 0 {
+			log.Infof("resource is owned by another object, defering deletion to parent resource(s): %v", o)
+			return nil
+		}
 
-	// Update PipelineRun with Result Annotations.
-	patch, err := annotation.Add(result.GetName(), record.GetName())
-	if err != nil {
-		log.Errorf("error adding Result annotations: %v", err)
-		return err
-	}
-	if _, err := r.pipelineclientset.TektonV1beta1().PipelineRuns(pr.GetNamespace()).Patch(pr.Name, types.MergePatchType, patch); err != nil {
-		log.Errorf("PipelineRun.Patch: %v", err)
-		return err
+		// We haven't hit the grace period yet - reenqueue the key for processing later.
+		if s := clock.Since(record.GetUpdatedTime().AsTime()); s < r.cfg.GetCompletedResourceGracePeriod() {
+			log.Infof("pipelinerun is not ready for deletion - grace period: %v, time since completion: %v", r.cfg.GetCompletedResourceGracePeriod(), s)
+			r.enqueue(pr.GetNamespacedName(), r.cfg.GetCompletedResourceGracePeriod())
+			return nil
+		}
+		log.Infof("deleting PipelineRun UID %s", pr.GetUID())
+		if err := r.pipelineclientset.TektonV1beta1().PipelineRuns(pr.GetNamespace()).Delete(pr.Name, &metav1.DeleteOptions{}); err != nil {
```

Comment:
This should handle `k8serrors.IsNotFound` as a no-op, in case we have dueling deletes.